metrics/resource_view.go

```diff
@@ -0,0 +1,324 @@
+			if view == viewToRemove {
+				toRemove = true
+				break
+			}
+		}
+		if !toRemove {
+			resourceViews.views[j] = view
+			j++
+		}
+	}
+	resourceViews.views = resourceViews.views[:j]
+}
+
+func setFactory(f ResourceExporterFactory) error {
+	if f == nil {
+		return errors.New("do not setFactory(nil)")
+	}
+
+	allMeters.lock.Lock()
+	defer allMeters.lock.Unlock()
+
+	allMeters.factory = f
+
+	var retErr error
+
+	for r, meter := range allMeters.meters {
+		e, err := f(resourceFromString(r))
+		if err != nil {
+			retErr = err
+			continue // Keep trying to clean up remaining Meters.
+		}
+		meter.m.UnregisterExporter(meter.e)
+		meter.m.RegisterExporter(e)
+		meter.e = e
+	}
+	return retErr
+}
+
+func flushResourceExporters() {
+	allMeters.lock.Lock()
+	defer allMeters.lock.Unlock()
+
+	for _, meter := range allMeters.meters {
+		flushGivenExporter(meter.e)
+	}
+}
+
+func meterExporterForResource(r *resource.Resource) *meterExporter {
+	key, ok := allMeters.resourceToKey[r]
+	if !ok {
+		key = resourceAsString(r)
+		allMeters.resourceToKey[r] = key
+	}
+
+	mE := allMeters.meters[key]
+	if mE == nil {
+		mE = &meterExporter{}
+		allMeters.meters[key] = mE
+	}
+	if mE.o != nil {
+		return mE
+	}
+	mE.m = view.NewMeter()
+	mE.m.SetResource(r)
+	mE.m.Start()
+	resourceViews.lock.Lock()
+	defer resourceViews.lock.Unlock()
+	// make a copy of views to avoid data races
+	viewsCopy := make([]*view.View, 0, len(resourceViews.views))
+	for _, v := range resourceViews.views {
+		c := *v
+		viewsCopy = append(viewsCopy, &c)
+	}
+	mE.m.Register(viewsCopy...)
+	mE.o = stats.WithRecorder(mE.m)
+	allMeters.meters[key] = mE
+	return mE
+}
+
+// optionForResource finds or creates a stats exporter for the resource, and
+// returns a stats.Option indicating which meter to record to.
+func optionForResource(r *resource.Resource) (stats.Options, error) {
+	allMeters.lock.Lock()
+	defer allMeters.lock.Unlock()
+
+	mE := meterExporterForResource(r)
+	if mE == nil {
+		return nil, fmt.Errorf("failed lookup for resource %v", r)
+	}
+
+	if mE.e != nil {
+		// Assume the exporter is already started.
+		return mE.o, nil
+	}
+
+	if allMeters.factory == nil {
+		if mE.o != nil {
+			// If we can't create exporters but we have a Meter, return that.
+			return mE.o, nil
+		}
+		return nil, fmt.Errorf("whoops, allMeters.factory is nil")
+	}
+	exporter, err := allMeters.factory(r)
+	if err != nil {
+		return nil, err
+	}
+
+	mE.m.RegisterExporter(exporter)
+	mE.e = exporter
+	return mE.o, nil
+}
+
+func resourceAsString(r *resource.Resource) string {
+	var s strings.Builder
+	l := len(r.Type)
+	kvs := make([]string, 0, len(r.Labels))
+	for k, v := range r.Labels {
+		l += len(k) + len(v) + 2
+		// We use byte values 1 and 2 to avoid colliding with valid resource labels
+		// and to make unpacking easy
+		kvs = append(kvs, fmt.Sprintf("\x01%s\x02%s", k, v))
+	}
+	s.Grow(l)
+	s.WriteString(r.Type)
+
+	sort.Strings(kvs)
+	s.WriteString(strings.Join(kvs, ""))
```

Comment:
I think this ruins the optimization we did in computing `l`.
Perhaps expand to a loop `for _, s := range kvs { s.WriteString(s)}`.

Reply:
Good point, except we don't want to name the loop variable "s".

Actually, I think I just want a C++ STL `map<string, string>` for `Resource.Labels`, but I don't get what I want.